### PR TITLE
Complex grid layout

### PR DESCRIPTION
**Pull request by t-bltg**
This is what I've come up with for https://github.com/FedeClaudi/Term.jl/pull/102#issuecomment-1157463510.

- [x] split `Grid` module from `Layout`;
- [x] reuse `Compositor` layout logic for complex grids;
- [x] hide placeholders by default unless the whole grid consists of placeholders;
- [x] find a solution for repeated elements as in https://matplotlib.org/3.5.0/tutorials/provisional/mosaic.html#string-short-hand;
- [x] ~~maybe use compositor for classic simple grids;~~ ==> return `Renderable(Compositor)` instead
- [x] ~~fractional size for `Compositor`.~~
- [x] swap `(width, height)` to `displaysize` convention;

That's a WIP for now (messy ?), but if you have comments ...

Example:
```julia
julia> using Term
julia> rens = [
    Panel(height = 5, width = 10),
    Panel(height = 5, width = 15),
    Panel(height = 5, width = 20),
    Panel(height = 10, width = 20),
    Panel(height = 15, width = 20),
]
julia> grid(rens, layout = :((_ * a) / (b * _ * c) / (d * e)), show_placeholder=true)
╲ ╲ ╲ ╲ ╲ ╭────────╮                         
 ╲ ╲ ╲ ╲ ╲│        │                         
╲ ╲ ╲ ╲ ╲ │        │                         
 ╲ ╲ ╲ ╲ ╲│        │                         
╲ ╲ ╲ ╲ ╲ ╰────────╯                         
╭─────────────╮╲ ╲ ╲ ╲ ╲ ╭──────────────────╮
│             │ ╲ ╲ ╲ ╲ ╲│                  │
│             │╲ ╲ ╲ ╲ ╲ │                  │
│             │ ╲ ╲ ╲ ╲ ╲│                  │
╰─────────────╯╲ ╲ ╲ ╲ ╲ ╰──────────────────╯
╭──────────────────╮╭──────────────────╮     
│                  ││                  │     
│                  ││                  │     
│                  ││                  │     
│                  ││                  │     
│                  ││                  │     
│                  ││                  │     
│                  ││                  │     
│                  ││                  │     
╰──────────────────╯│                  │     
                    │                  │     
                    │                  │     
                    │                  │     
                    │                  │     
                    ╰──────────────────╯     


```

### CODE REVIEW

**Comment by t-bltg** [comment]
Something to discuss: in `Term.jl` sizes are expressed in terms of `(width, height)` which maps to `(columns, rows)` which is not very intuitive: should we instead use the swapped `(height, width)` == `(rows, cols)` convention (this is what we use in `UnicodePlots` and what is used in `Base` e.g. for `displaysize` ?

File: src/grid.jl [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    (isnothing(layout) && isnothing(rens)) && error("Grid: layout or aspect required")
+
+    if !isnothing(rens) && layout isa Expr
+        n, kw = 0, Dict()
+        sz = size(Measure(first(rens).segments))
+        for e in get_elements_and_sizes(layout; placeholder_size = sz)
+            (nm = e.args[1]) === :_ && continue
+            kw[nm] = rens[n += 1]
+        end
+        return Compositor(layout; placeholder_size = sz, check = false, pairs(kw)...)
```
Comment (FedeClaudi) [inline src/grid.jl:50]: This should be `Renderable(Compositor(...))`  to keep it consistent with the other behavior where `grid` returns a renderable

**Comment by FedeClaudi** [comment]
Looks great, it's going to be very useful!
How do you ignore `_`?


---
It might be a good opportunity to implement in `Compositor` the option to specify renderables with a "fractional" size.

```julia
layout = :(
    A(.7, 1.0) * B(.3, 1.0)
)
Compositor(layout)
```

should create a layout in which `A` takes up 70% of the console width (or an user-provided `width`). It should be easy, just multiply the fractional value by the width when creating the `LayoutElement`. It might just require a bit a care while parsing the `layout` expression.

**Comment by FedeClaudi** [comment]
> Now will be a good time to change this, before a breaking change.

Indeed it would. Would it be just a matter of replacing the order, or would you call them rows/columns?

File: src/grid.jl (hunk 1 above)
Comment (t-bltg) [reply src/grid.jl]: Thanks, fixed.

**Comment by t-bltg** [comment]
> How do you ignore _

https://github.com/FedeClaudi/Term.jl/pull/113/files#diff-864bc1ab10d7005406c446aed073e078149d74a75995a0b1b103a9b09220c73aR83

> Would it be just a matter of replacing the order

Yes, I think swapping the order will be ok. I just wanted to confirm that you are ok with this change. I can do that now.

> Did you disable the show placeholders stuff in grid btw? I can do that otherwise

Yes, done.

> It might be a good opportunity to implement in Compositor the option to specify renderables with a "fractional" size.

Ha ok, I thought it was already implemented. Will add to the TODO list.

**Comment by FedeClaudi** [comment]
yeah sure let's do the swap.

No it's not, but I didn't mean you have to do it! I can do that too, you've already dedicated so much time to Term!

**Comment by t-bltg** [comment]
> I can do that too, you've already dedicated so much time to Term!

OK, let's leave that to another PR ;)

**Comment by FedeClaudi** [comment]
sounds good!

**Comment by FedeClaudi** [comment]
> In the following example, placeholders are correctly hidden in the terminal (but appearing on github rendering):

It's interesting that it adds two placeholders but there's gaps elsewhere. Are the gaps filled ? 
As in, if you stack two such grids next to each other do they look fine?

They should be since that's what the stacking code does, but it would be good to make sure.

**Comment by t-bltg** [comment]
The swap is now done, will all tests passing, and I've tried to carefully swap things in the docs (e.g. `PlaceHolder`, `Spacer`, but I think this is worth a second look.

**Comment by t-bltg** [comment]
> It's interesting that it adds two placeholders but there's gaps elsewhere. Are the gaps filled ?
As in, if you stack two such grids next to each other do they look fine?

Yep, that's worth enhancing the algorithm here.

**Comment by FedeClaudi** [comment]
I didn't know about `something`, you're teaching me a lot!

> The swap is now done, will all tests passing, and I've tried to carefully swap things in the docs (e.g. PlaceHolder, Spacer), but I think this is worth a second look.

Thank you so much! Yeah I also need to change the section on `grid` to reflect the new changes

**Comment by t-bltg** [comment]
> Yeah I also need to change the section on grid to reflect the new changes

I've added a note in the docs about the `layout as Expr` for `grid`.

**Comment by FedeClaudi** [comment]
First of all, that's awesome. It's crazy how quickly you've pulled that together. 
What I mean is, if you do:

```julia
g1 = grid(rens, layout = :((_ * a) / (b * _ * c) / (d * e)), show_placeholder=true)
g2 = grid(rens, layout = :((_ * a) / (b * _ * c) / (d * e)), show_placeholder=true)
```
does it look fine?  As in two grids side by side, instead of all mixed up. 



> I'm a bit out of ideas for now on how to enhance the layout thing.

What are you stuck on? How can I help?

**Comment by t-bltg** [comment]
> g1 = grid(rens, layout = :((_ * a) / (b * _ * c) / (d * e)), show_placeholder=true)
g2 = grid(rens, layout = :((_ * a) / (b * _ * c) / (d * e)), show_placeholder=true)
g1 * g2  # stack em

Yeah, looks ok in terms of addes spaces on the right.

```julia
╲ ╲ ╲ ╲ ╲ ╭────────╮                         ╲ ╲ ╲ ╲ ╲ ╭────────╮                         
 ╲ ╲ ╲ ╲ ╲│        │                          ╲ ╲ ╲ ╲ ╲│        │                         
╲ ╲ ╲ ╲ ╲ │        │                         ╲ ╲ ╲ ╲ ╲ │        │                         
 ╲ ╲ ╲ ╲ ╲│        │                          ╲ ╲ ╲ ╲ ╲│        │                         
╲ ╲ ╲ ╲ ╲ ╰────────╯                         ╲ ╲ ╲ ╲ ╲ ╰────────╯                         
╭─────────────╮╲ ╲ ╲ ╲ ╲ ╭──────────────────╮╭─────────────╮╲ ╲ ╲ ╲ ╲ ╭──────────────────╮
│             │ ╲ ╲ ╲ ╲ ╲│                  ││             │ ╲ ╲ ╲ ╲ ╲│                  │
│             │╲ ╲ ╲ ╲ ╲ │                  ││             │╲ ╲ ╲ ╲ ╲ │                  │
│             │ ╲ ╲ ╲ ╲ ╲│                  ││             │ ╲ ╲ ╲ ╲ ╲│                  │
╰─────────────╯╲ ╲ ╲ ╲ ╲ ╰──────────────────╯╰─────────────╯╲ ╲ ╲ ╲ ╲ ╰──────────────────╯
╭──────────────────╮╭──────────────────╮     ╭──────────────────╮╭──────────────────╮     
│                  ││                  │     │                  ││                  │     
│                  ││                  │     │                  ││                  │     
│                  ││                  │     │                  ││                  │     
│                  ││                  │     │                  ││                  │     
│                  ││                  │     │                  ││                  │     
│                  ││                  │     │                  ││                  │     
│                  ││                  │     │                  ││                  │     
│                  ││                  │     │                  ││                  │     
╰──────────────────╯│                  │     ╰──────────────────╯│                  │     
                    │                  │                         │                  │     
                    │                  │                         │                  │     
                    │                  │                         │                  │     
                    │                  │                         │                  │     
                    ╰──────────────────╯                         ╰──────────────────╯     
```

**Comment by FedeClaudi** [comment]
excellent.

I think what you have already is very powerful:
1. if user just want a regular grid they can use `grid` as it was before
2. if they want a grid with "gaps" they can use the new grid
3. if they want something more complicated they can use `Compositor`

**Comment by t-bltg** [comment]
> What are you stuck on? How can I help?

Pretty minor, but I would find it easier to use e.g. `layout = :((_ * ⬓) / (⬓ * _ * ⬓) / (⬓ * ⬓))` (same symbol instead of `a, b, ...`), but my tentative generating unique symbols for the compositor dict failed. Anyway not really important.

**Comment by t-bltg** [comment]
I'm sure there will be more corner cases to fix, or validate inputs cleanly (instead of erroring with invalid bounds), but yeah it's usable now.

**Comment by FedeClaudi** [comment]
> but my tentative generating unique symbols for the compositor dict failed. Anyway not really important.

In compositors elements with the same name are considered to be the same renderable shown in two places (e.g. a vertical line). So that might be causing you issues perhaps?

Yeah the matplotlib version would be the best, but I've found it very hard to go from a string to a layout expression. Mostly because it is not a 1-1 mapping, you could use many different expressions to get the same layout... it's hard. Perhaps for the future!

**Comment by t-bltg** [comment]
This now works for repeated elements in layout !

```julia
julia> grid(rens, layout = :((_ * a) / (b * _ * c * c) / (d * e)))
╲ ╲ ╲ ╲ ╲ ╭────────╮                                             
 ╲ ╲ ╲ ╲ ╲│        │                                             
╲ ╲ ╲ ╲ ╲ │        │                                             
 ╲ ╲ ╲ ╲ ╲│        │                                             
╲ ╲ ╲ ╲ ╲ ╰────────╯                                             
╭─────────────╮╲ ╲ ╲ ╲ ╲ ╭──────────────────╮╭──────────────────╮
│             │ ╲ ╲ ╲ ╲ ╲│                  ││                  │
│             │╲ ╲ ╲ ╲ ╲ │                  ││                  │
│             │ ╲ ╲ ╲ ╲ ╲│                  ││                  │
╰─────────────╯╲ ╲ ╲ ╲ ╲ ╰──────────────────╯╰──────────────────╯
╭──────────────────╮╭──────────────────╮                         
│                  ││                  │                         
│                  ││                  │                         
│                  ││                  │                         
│                  ││                  │                         
│                  ││                  │                         
│                  ││                  │                         
│                  ││                  │                         
│                  ││                  │                         
╰──────────────────╯│                  │                         
                    │                  │                         
                    │                  │                         
                    │                  │                         
                    │                  │                         
                    ╰──────────────────╯                         

```

**Comment by FedeClaudi** [comment]
let me know when you're happy to merge!

**Comment by t-bltg** [comment]
Ci green, good to go ;)

**Comment by FedeClaudi** [comment]
squash and merge ;)

**Comment by t-bltg** [comment]
Grid in action: https://github.com/JuliaPlots/UnicodePlots.jl#unicodeplots.

**Pull request merged**